test/composableStableJoin.integration.test.ts

```diff
@@ -0,0 +1,283 @@
+        // setup mock api
+        const api = new MockApi();
+
+        // get pool state from api
+        const poolInput = await api.getPool(poolId);
+
+        const client = createTestClient({
+            mode: 'anvil',
+            chain: CHAINS[chainId],
+            transport: http(rpcUrl),
+        })
+            .extend(publicActions)
+            .extend(walletActions);
+
+        txInput = {
+            client,
+            poolJoin: new PoolJoin(),
+            slippage: Slippage.fromPercentage('1'), // 1%
+            poolStateInput: poolInput,
+            testAddress: '0x10a19e7ee7d7f8a52822f6817de8ea18204f2e4f', // Balancer DAO Multisig
+            joinInput: {} as JoinInput,
+        };
+
+        // setup BPT token
+        bptToken = new Token(chainId, poolInput.address, 18, 'BPT');
+    });
+
+    beforeEach(async () => {
+        await forkSetup(
+            txInput.client,
+            txInput.testAddress,
+            [...txInput.poolStateInput.tokens.map((t) => t.address)],
+            [0, 0, 3],
+            [
+                ...txInput.poolStateInput.tokens.map((t) =>
+                    parseUnits('100', t.decimals),
+                ),
+            ],
+        );
+    });
+
+    describe('unbalanced join', () => {
+        let input: Omit<UnbalancedJoinInput, 'amountsIn'>;
+        let amountsIn: TokenAmount[];
+        beforeAll(() => {
+            const bptIndex = txInput.poolStateInput.tokens.findIndex(
+                (t) => t.address === txInput.poolStateInput.address,
+            );
+            const poolTokensWithoutBpt = txInput.poolStateInput.tokens
+                .map((t) => new Token(chainId, t.address, t.decimals))
+                .filter((_, index) => index !== bptIndex);
+
+            amountsIn = poolTokensWithoutBpt.map((t) =>
+                TokenAmount.fromHumanAmount(t, '1'),
+            );
+            input = {
+                chainId,
+                rpcUrl,
+                kind: JoinKind.Unbalanced,
+            };
+        });
+        test('token inputs', async () => {
+            const joinInput = {
+                ...input,
+                amountsIn: [...amountsIn.splice(0, 1)],
+            };
+            const joinResult = await doJoin({
+                ...txInput,
+                joinInput,
+            });
+            assertUnbalancedJoin(
+                txInput.client.chain?.id as number,
+                txInput.poolStateInput,
+                joinInput,
+                joinResult,
+                txInput.slippage,
+            );
+        });
+
+        test('with native', async () => {
+            const joinInput = {
+                ...input,
+                amountsIn,
+                useNativeAssetAsWrappedAmountIn: true,
+            };
+            const joinResult = await doJoin({
+                ...txInput,
+                joinInput: {
+                    ...joinInput,
+                    useNativeAssetAsWrappedAmountIn: true,
```

Comment:
You should be able to remove this, since you're already adding it to joinInput above.